prov/usnic/src/usdf_dgram.c

```diff
@@ -468,62 +499,121 @@ usdf_dgram_prefix_send(struct fid_ep *fep, const void *buf, size_t len,
+		sizeof(struct ether_header) -
+		sizeof(struct iphdr)) + len);
+	hdr->uh_udp.source =
+		qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
+
+	memcpy(send_iov, iov, sizeof(struct iovec) * count);
+	send_iov[0].iov_base = hdr;
+	send_iov[0].iov_len += sizeof(*hdr);
+
+	last_post = _usd_post_send_iov(wq, send_iov, count, cq_entry);
+	info = &wq->uwq_post_info[last_post];
+	info->wp_context = context;
+	info->wp_len = len;
+
+	return 0;
+}
+
+ssize_t
 usdf_dgram_prefix_sendv(struct fid_ep *fep, const struct iovec *iov, void **desc,
 		size_t count, fi_addr_t dest_addr, void *context)
 {
 	struct usdf_ep *ep;
 	struct usd_dest *dest;
-	struct usd_wq *wq;
- 	struct usd_qp_impl *qp;
-	struct usd_udp_hdr *hdr;
-	uint32_t last_post;
-	struct usd_wq_post_info *info;
-	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
-	size_t len;
-	unsigned i;
+
+	uint8_t completion;
+	size_t send_size = sizeof(struct usd_udp_hdr);
 
 	ep = ep_ftou(fep);
 	dest = (struct usd_dest *)(uintptr_t) dest_addr;
 
-	len = 0;
-	for (i = 0; i < count; i++) {
-		len += iov[i].iov_len;
+	for (size_t i = 0; i < count; i++) {
+		send_size += iov[i].iov_len;
 	}
 
-	if (len + sizeof(struct usd_udp_hdr) > USD_SEND_MAX_COPY) {
-		qp = to_qpi(ep->e.dg.ep_qp);
-		wq = &qp->uq_wq;
-		hdr = (struct usd_udp_hdr *) iov[0].iov_base - 1;
-		memcpy(hdr, &dest->ds_dest.ds_udp.u_hdr, sizeof(*hdr));
-
-		/* adjust lengths and insert source port */
-		hdr->uh_ip.tot_len = htons(len + sizeof(struct usd_udp_hdr) -
-			sizeof(struct ether_header));
-		hdr->uh_udp.len = htons((sizeof(struct usd_udp_hdr) -
-			sizeof(struct ether_header) -
-			sizeof(struct iphdr)) + len);
-		hdr->uh_udp.source =
-			qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
-
-		memcpy(send_iov, iov, sizeof(struct iovec) * count);
-		send_iov[0].iov_base = hdr;
-		send_iov[0].iov_len += sizeof(*hdr);
-
-		last_post = _usd_post_send_iov(wq, send_iov, count, 1);
-		info = &wq->uwq_post_info[last_post];
-		info->wp_context = context;
-		info->wp_len = len;
-	} else {
-		_usdf_dgram_send_iov_copy(ep, dest, iov, count, context);
+	completion = ep->ep_tx_dflt_signal_comp ||
+		(ep->e.dg.tx_op_flags & FI_COMPLETION);
+
+	if (send_size <= USD_SEND_MAX_COPY) {
+		return _usdf_dgram_send_iov_copy_prefix(ep, dest, iov, count,
+			context, completion);
+	} else if (ep->e.dg.tx_op_flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA, "max inject size exceeded: %zu\n",
+			send_size);
+		return -FI_EINVAL;
 	}
-	return 0;
+
+	return _usdf_dgram_send_iov_prefix(ep, dest, iov, count, context,
+		completion);
 }
 
 ssize_t
 usdf_dgram_prefix_sendmsg(struct fid_ep *fep, const struct fi_msg *msg, uint64_t flags)
 {
-	return usdf_dgram_prefix_sendv(fep, msg->msg_iov, msg->desc, msg->iov_count,
-				(fi_addr_t)msg->addr, msg->context);
+	struct usdf_ep *ep;
+	struct usd_dest *dest;
+	uint8_t completion;
+	size_t send_size = sizeof(struct usd_udp_hdr);
+
+	ep = ep_ftou(fep);
+	dest = (struct usd_dest *)(uintptr_t)(fi_addr_t) msg->addr;
```

Comment:
Is the `(fi_addr_t)` cast actually needed?  Isn't that already `msg->addr`'s type?